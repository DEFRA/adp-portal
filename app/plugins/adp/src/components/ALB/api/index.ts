export { ArmsLengthBodyClient } from './AlbClient'
export type {ArmsLengthBodyApi} from './AlbApi'
