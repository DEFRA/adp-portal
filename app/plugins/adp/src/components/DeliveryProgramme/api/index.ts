export { DeliveryProgrammeClient } from './DeliveryProgrammeClient';
export type { DeliveryProgrammeApi } from './DeliveryProgrammeApi';
