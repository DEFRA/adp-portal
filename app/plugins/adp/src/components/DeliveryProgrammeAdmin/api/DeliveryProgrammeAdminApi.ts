import type { DeliveryProgrammeAdmin } from '@internal/plugin-adp-common';
import { createApiRef } from '@backstage/core-plugin-api';

export const deliveryProgrammeAdminApiRef =
  createApiRef<DeliveryProgrammeAdminApi>({
    id: 'plugin.adp.deliveryprogrammeadminapi',
  });

export interface DeliveryProgrammeAdminApi {
  getAll(): Promise<DeliveryProgrammeAdmin[]>;
  getByDeliveryProgrammeId(
    deliveryProgrammeId: string,
  ): Promise<DeliveryProgrammeAdmin[]>;
  create(
    deliveryProgrammeId: string,
    userCatalogName: string,
    groupEntityRef: string,
  ): Promise<DeliveryProgrammeAdmin>;
  delete(aadEntityRefId: string, deliveryProgrammeId: string): void;
}
