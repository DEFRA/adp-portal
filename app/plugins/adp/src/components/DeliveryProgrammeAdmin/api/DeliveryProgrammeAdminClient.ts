import type { DeliveryProgrammeAdmin } from '@internal/plugin-adp-common';
import type { DiscoveryApi, FetchApi } from '@backstage/core-plugin-api';
import { ResponseError } from '@backstage/errors';
import type { DeliveryProgrammeAdminApi } from './DeliveryProgrammeAdminApi';

export class DeliveryProgrammeAdminClient
  implements DeliveryProgrammeAdminApi
{
  private discoveryApi: DiscoveryApi;
  private fetchApi: FetchApi;

  constructor(discoveryApi: DiscoveryApi, fetchApi: FetchApi) {
    this.discoveryApi = discoveryApi;
    this.fetchApi = fetchApi;
  }

  public async getByDeliveryProgrammeId(
    deliveryProgrammeId: string,
  ): Promise<DeliveryProgrammeAdmin[]> {
    const baseUrl = await this.getBaseUrl();
    const url = `${baseUrl}/deliveryProgrammeAdmins/${deliveryProgrammeId}`;

    const response = await this.fetchApi.fetch(url);

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    const deliveryProgrammeAdmins =
      (await response.json()) as DeliveryProgrammeAdmin[];

    return deliveryProgrammeAdmins;
  }

  async getAll(): Promise<DeliveryProgrammeAdmin[]> {
    const baseUrl = await this.getBaseUrl();
    const url = `${baseUrl}/deliveryProgrammeAdmins/`;

    const response = await this.fetchApi.fetch(url);

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    const deliveryProgrammeAdmins =
      (await response.json()) as DeliveryProgrammeAdmin[];

    return deliveryProgrammeAdmins;
  }

  async create(
    deliveryProgrammeId: string,
    aadEntityRefIds: string[],
  ): Promise<DeliveryProgrammeAdmin[]> {
    const baseUrl = await this.getBaseUrl();
    const url = `${baseUrl}/deliveryProgrammeAdmin/${deliveryProgrammeId}`;

    const body = {
      aadEntityRefIds: aadEntityRefIds
    };

    const response = await this.fetchApi.fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }

    const deliveryProgrammeAdmins =
      (await response.json()) as DeliveryProgrammeAdmin[];

    return deliveryProgrammeAdmins;
  }

  async delete(
    aadEntityRefId: string,
    deliveryProgrammeId: string,
  ) {
    const baseUrl = await this.getBaseUrl();
    const url = `${baseUrl}/deliveryProgrammeAdmin`;

    const body = {
      aadEntityRefId: aadEntityRefId,
      deliveryProgrammeId: deliveryProgrammeId
    };

    const response = await this.fetchApi.fetch(url, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      throw await ResponseError.fromResponse(response);
    }
  }

  private async getBaseUrl(): Promise<string> {
    return `${await this.discoveryApi.getBaseUrl('adp')}`;
  }
}
