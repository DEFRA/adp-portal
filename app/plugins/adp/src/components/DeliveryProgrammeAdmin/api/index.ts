export { deliveryProgrammeAdminApiRef } from './DeliveryProgrammeAdminApi';
export { DeliveryProgrammeAdminApiClient } from './DeliveryProgrammeAdminClient';
