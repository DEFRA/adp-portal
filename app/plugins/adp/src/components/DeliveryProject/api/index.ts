export * from './DeliveryProjectClient';
export * from './DeliveryProjectApi';
