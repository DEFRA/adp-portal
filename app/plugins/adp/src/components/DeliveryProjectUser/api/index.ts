export {
  deliveryProjectUserApiRef,
  type DeliveryProjectUserApi,
} from './DeliveryProjectUserApi';
export { DeliveryProjectUserClient } from './DeliveryProjectUserClient';
