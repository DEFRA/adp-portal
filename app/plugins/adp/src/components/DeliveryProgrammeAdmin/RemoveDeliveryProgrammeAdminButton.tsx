import { alertApiRef, useApi } from '@backstage/core-plugin-api';
import { Button } from '@material-ui/core';
import React, { useState } from 'react';
import { deliveryProgrammeAdminApiRef } from './api';
import { usePermission } from '@backstage/plugin-permission-react';
import type { DeliveryProgrammeAdmin } from '@internal/plugin-adp-common';
import { deliveryProgrammeAdminDeletePermission } from '@internal/plugin-adp-common';
import { ConfirmationDialog } from '../../utils';

export type RemoveDeliveryProgrammeAdminButtonProps = Readonly<
  Omit<Parameters<typeof Button>[0], 'onClick'> & {
    deliveryProgrammeAdmin: DeliveryProgrammeAdmin;
    entityRef: string;
    onRemoved?: () => void;
  }
>;

export function RemoveDeliveryProgrammeAdminButton({
  deliveryProgrammeAdmin,
  entityRef,
  onRemoved,
  ...buttonProps
}: RemoveDeliveryProgrammeAdminButtonProps) {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const alertApi = useApi(alertApiRef);
  const client = useApi(deliveryProgrammeAdminApiRef);
  const { allowed: canRemoveProgrammeAdmin } = usePermission({
    permission: deliveryProgrammeAdminDeletePermission,
    resourceRef: entityRef,
  });

  if (!canRemoveProgrammeAdmin) return null;

  async function handleDelete() {
    await client.delete(deliveryProgrammeAdmin.id, entityRef);
    onRemoved?.();
    alertApi.post({
      message: `Removed ${deliveryProgrammeAdmin.name} from this delivery programme`,
      display: 'transient',
      severity: 'success',
    });
  }

  return (
    <>
      <Button
        {...buttonProps}
        onClick={() => setIsModalOpen(true)}
        data-testid="remove-delivery-programme-admin-button"
      >
        Remove
      </Button>

      {isModalOpen && (
        <ConfirmationDialog
          title={`Remove ${deliveryProgrammeAdmin.name}?`}
          content="Are you sure you want to remove this user? The user will no longer be able to perform certain actions on the ADP portal. You can re-add the user after removing them."
          confirm="Remove"
          confirmColour="secondary"
          completed={submitted => {
            setIsModalOpen(false);
            if (submitted) onRemoved?.();
          }}
          submit={handleDelete}
        />
      )}
    </>
  );
}
