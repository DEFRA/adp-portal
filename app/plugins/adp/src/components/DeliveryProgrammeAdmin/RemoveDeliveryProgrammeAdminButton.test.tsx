import { alertApiRef, type AlertApi } from '@backstage/core-plugin-api';
import React from 'react';
import type { RemoveDeliveryProgrammeAdminButtonProps } from './RemoveDeliveryProgrammeAdminButton';
import { RemoveDeliveryProgrammeAdminButton } from './RemoveDeliveryProgrammeAdminButton';
import { render, waitFor } from '@testing-library/react';
import { TestApiProvider } from '@backstage/test-utils';
import type { DeliveryProgrammeAdminApi } from './api';
import { deliveryProgrammeAdminApiRef } from './api';
import { SnapshotFriendlyStylesProvider } from '../../utils';
import type * as PluginPermissionReactModule from '@backstage/plugin-permission-react';

function setup() {
  const mockAlertApi: jest.Mocked<AlertApi> = {
    alert$: jest.fn(),
    post: jest.fn(),
  };
  const mockProgrammeAdminApi: jest.Mocked<DeliveryProgrammeAdminApi> = {
    create: jest.fn(),
    delete: jest.fn(),
    getAll: jest.fn(),
    getByDeliveryProgrammeId: jest.fn(),
  };

  return {
    mockAlertApi,
    mockProgrammeAdminApi,
    async renderComponent(props: RemoveDeliveryProgrammeAdminButtonProps) {
      const result = render(
        <TestApiProvider
          apis={[
            [alertApiRef, mockAlertApi],
            [deliveryProgrammeAdminApiRef, mockProgrammeAdminApi],
          ]}
        >
          <SnapshotFriendlyStylesProvider>
            <RemoveDeliveryProgrammeAdminButton {...props} />
          </SnapshotFriendlyStylesProvider>
        </TestApiProvider>,
      );
      await waitFor(() => expect(result.baseElement).not.toBeEmptyDOMElement());
      return { result };
    },
  };
}

const deliveryProgrammeAdmin = {
  aad_entity_ref_id: '123',
  delivery_programme_id: '123',
  email: `test-programme-admin@test.com`,
  id: '123',
  name: 'Test Delivery Programme Admin',
  updated_at: new Date(0),
};

const usePermission: jest.MockedFn<
  typeof PluginPermissionReactModule.usePermission
> = jest.fn();

jest.mock(
  '@backstage/plugin-permission-react',
  () =>
    ({
      get usePermission() {
        return usePermission;
      },
      get IdentityPermissionApi(): never {
        throw new Error('Not mocked');
      },
      get PermissionedRoute(): never {
        throw new Error('Not mocked');
      },
      get RequirePermission(): never {
        throw new Error('Not mocked');
      },
      get permissionApiRef(): never {
        throw new Error('Not mocked');
      },
    }) satisfies typeof PluginPermissionReactModule,
);

describe('RemoveDeliveryProgrammeAdminButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not render when the user is not allowed to remove delivery programme admins', async () => {
    const { mockAlertApi, mockProgrammeAdminApi, renderComponent } = setup();
    usePermission.mockReturnValue({ allowed: false, loading: false });

    const { result } = await renderComponent({
      content: 'Test button',
      deliveryProgrammeAdmin: deliveryProgrammeAdmin,
      entityRef: 'programme-group-123',
    });

    expect(result.baseElement).toMatchSnapshot();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.create).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.delete).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.getAll).not.toHaveBeenCalled();
    expect(
      mockProgrammeAdminApi.getByDeliveryProgrammeId,
    ).not.toHaveBeenCalled();
  });

  it('should only render a button initially', async () => {
    const { renderComponent, mockAlertApi, mockProgrammeAdminApi } = setup();
    usePermission.mockReturnValue({ allowed: true, loading: false });

    const { result } = await renderComponent({
      content: 'Test button',
      deliveryProgrammeAdmin: deliveryProgrammeAdmin,
      entityRef: 'programme-group-123',
    });

    expect(result.baseElement).toMatchSnapshot();
    expect(mockAlertApi.alert$).not.toHaveBeenCalled();
    expect(mockAlertApi.post).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.create).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.delete).not.toHaveBeenCalled();
    expect(mockProgrammeAdminApi.getAll).not.toHaveBeenCalled();
    expect(
      mockProgrammeAdminApi.getByDeliveryProgrammeId,
    ).not.toHaveBeenCalled();
  });
});
