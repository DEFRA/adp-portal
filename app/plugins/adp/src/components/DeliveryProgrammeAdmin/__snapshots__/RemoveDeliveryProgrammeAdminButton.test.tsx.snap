// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveDeliveryProgrammeAdminButton should call onRemoved when the dialog is closed via the Remove action: After cancel 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should call onRemoved when the dialog is closed via the Remove action: After complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should call onRemoved when the dialog is closed via the Remove action: Before complete 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      Test dialog
    </span>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should call the API when submitting 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      Test dialog
    </span>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should close the dialog via the Cancel action: After cancel 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should close the dialog via the Cancel action: Before cancel 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      Test dialog
    </span>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should not render when the user is not allowed to remove delivery programme admins 1`] = `
<body>
  <div />
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should only render a button initially 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`RemoveDeliveryProgrammeAdminButton should render the dialog when the button is clicked 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      content="Test button"
      data-testid="remove-delivery-programme-admin-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Remove
      </span>
      <span
        class="MuiTouchRipple-root"
      >
        <span
          class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <span>
      Test dialog
    </span>
  </div>
</body>
`;
